pkg/util/testutil/docker/config.go

```diff
@@ -0,0 +1,182 @@
+package docker
+
+import (
+	"fmt"
+	"regexp"
+	"time"
+)
+
+const (
+	// DefaultTimeout is the default timeout for running a server.
+	DefaultTimeout = time.Minute
+
+	// DefaultRetries is the default number of retries for starting a container/s.
+	DefaultRetries = 3
+)
+
+type commandType string
+
+const (
+	dockerCommand commandType = "docker"
+	// we are using old v1 docker-compose command because our CI doesn't support docker cli v2 yet
+	composeCommand commandType = "docker-compose"
+	runCommand     commandType = "run"
+	removeCommand  commandType = "rm"
+)
+
+type subCommandType int
+
+const (
+	start = iota
+	kill
+)
+
+// Compile-time interface compliance check
+var _ LifecycleConfig = (*runConfig)(nil)
+var _ LifecycleConfig = (*composeConfig)(nil)
+
+// LifecycleConfig is an interface for the common configuration of a container lifecycle.
+type LifecycleConfig interface {
+	Timeout() time.Duration
+	Retries() int
+	LogPattern() *regexp.Regexp
+	Env() []string
+	Name() string
+	command() string
+	commandArgs(t subCommandType) []string
+}
+
+// Timeout returns the timeout to be used when running a container/s
+func (b baseConfig) Timeout() time.Duration {
+	return b.timeout
+}
+
+// Retries returns the number of retries to be used when trying to start the container/s
+func (b baseConfig) Retries() int {
+	return b.retries
+}
+
+// LogPattern returns the regex pattern to match logs for readiness
+func (b baseConfig) LogPattern() *regexp.Regexp {
+	return b.logPattern
+}
+
+// Env returns the environment variables to set for the container/s
+func (b baseConfig) Env() []string {
+	return b.env
+}
+
+// Name returns the name of the docker container or a friendly name for the docker-compose setup
+func (b baseConfig) Name() string {
+	return b.name
+}
+
+// BaseConfig contains shared configurations for both Docker and Docker Compose.
+type baseConfig struct {
+	name       string         // Container name for docker or an alias for docker-compose
+	timeout    time.Duration  // Timeout for the entire operation.
+	retries    int            // Number of retries for starting.
+	logPattern *regexp.Regexp // Regex pattern to match logs for readiness.
+	env        []string       // Environment variables to set.
+}
+
+// runConfig contains specific configurations for Docker containers, embedding BaseConfig.
+type runConfig struct {
+	baseConfig                   // Embed general configuration.
+	ImageName  string            // Docker image to use.
+	Binary     string            // Binary to run inside the container.
+	BinaryArgs []string          // Arguments for the binary.
+	Mounts     map[string]string // Mounts (host path -> container path).
+}
+
+func (r runConfig) command() string {
+	return string(dockerCommand)
+}
+
+func (r runConfig) commandArgs(t subCommandType) []string {
+	var args []string
+	switch t {
+	case start:
+		args = []string{string(runCommand), "--rm"}
```

Comment:
I do recommend on adding inline comments to clarify what you're doing
For instance why do we need `--rm`?